File: e2e/spikes/asr-1.0/doubao-tts.mjs
// doubao seed-tts-2.0 v3 双向流式客户端(spike 版;协议移植自 owner repo-demo-recorder
// scripts/add-tts-narration.mjs,07 D5 指定的复用来源;砍掉录屏混音,加首包延迟统计)。

import { randomUUID } from "node:crypto";
import { WebSocket } from "ws";

const MSG_TYPE = { FullClientRequest: 0b0001, FullServerResponse: 0b1001, AudioOnlyServer: 0b1011, Error: 0b1111 };
const FLAG = { NoSeq: 0, PositiveSeq: 0b1, NegativeSeq: 0b11, WithEvent: 0b100 };
const EVENT = {
  StartConnection: 1,
  FinishConnection: 2,
  ConnectionStarted: 50,
  ConnectionFailed: 51,
  ConnectionFinished: 52,
  StartSession: 100,
  FinishSession: 102,
  SessionStarted: 150,
  SessionFinished: 152,
  SessionFailed: 153,
  TaskRequest: 200,
  TTSSentenceStart: 350,
  TTSSentenceEnd: 351,
  TTSResponse: 352
};

const writeInt32 = (v) => { const b = Buffer.alloc(4); b.writeInt32BE(v); return b; };
const writeUInt32 = (v) => { const b = Buffer.alloc(4); b.writeUInt32BE(v); return b; };
const writeString = (v) => { const b = Buffer.from(v || "", "utf8"); return Buffer.concat([writeUInt32(b.length), b]); };
const skipSession = (e) => [EVENT.StartConnection, EVENT.FinishConnection, EVENT.ConnectionStarted, EVENT.ConnectionFailed, EVENT.ConnectionFinished].includes(e);

function encode({ event, sessionId = "", payload = "{}" }) {
  const pb = Buffer.from(payload, "utf8");
  const chunks = [Buffer.from([0x11, (MSG_TYPE.FullClientRequest << 4) | FLAG.WithEvent, 0x10, 0x00]), writeInt32(event)];
  if (!skipSession(event)) chunks.push(writeString(sessionId));
  chunks.push(writeUInt32(pb.length), pb);
  return Buffer.concat(chunks);
}

function decode(data) {
  const buf = Buffer.isBuffer(data) ? data : Buffer.from(data);
  const headerSize = (buf[0] & 0x0f) * 4;
  const type = buf[1] >> 4;
  const flag = buf[1] & 0x0f;
  let off = headerSize;
  const readInt32 = () => { const v = buf.readInt32BE(off); off += 4; return v; };
  const readUInt32 = () => { const v = buf.readUInt32BE(off); off += 4; return v; };
  const readStr = () => { const n = readUInt32(); const v = buf.subarray(off, off + n).toString("utf8"); off += n; return v; };
  if ([MSG_TYPE.FullClientRequest, MSG_TYPE.FullServerResponse, MSG_TYPE.AudioOnlyServer].includes(type)) {
    if (flag === FLAG.PositiveSeq || flag === FLAG.NegativeSeq) readInt32();
  } else if (type === MSG_TYPE.Error) {
    readUInt32();
  }
  let event = 0;
  if (flag === FLAG.WithEvent) {
    event = readInt32();
    if (!skipSession(event)) readStr();
    if ([EVENT.ConnectionStarted, EVENT.ConnectionFailed, EVENT.ConnectionFinished].includes(event)) readStr();
  }
  const payloadSize = off + 4 <= buf.length ? readUInt32() : 0;
  const payload = payloadSize > 0 ? buf.subarray(off, off + payloadSize) : Buffer.alloc(0);
  return { type, event, payload };
}

/**
 * 合成一句(每句独立连接;返回 mp3 Buffer + 首包延迟)。
 * speechRate:[-50, 100],0=正常。
 */
export async function synthesize(text, { apiKey, resourceId = "seed-tts-2.0", model = "seed-tts-2.0-expressive", voice = "zh_female_jitangmei_uranus_bigtts", sampleRate = 24000, speechRate = 0, endpoint = "wss://openspeech.bytedance.com/api/v3/tts/bidirection" }) {
  const ws = new WebSocket(endpoint, {
    headers: { "X-Api-Key": apiKey, "X-Api-Resource-Id": resourceId, "X-Api-Connect-Id": randomUUID() }
  });
  const queue = [];
  const waiters = [];
  let failed = null;
  ws.on("message", (d) => { const m = decode(d); const w = waiters.shift(); w ? w.resolve(m) : queue.push(m); });
  ws.on("error", (e) => { failed = e; while (waiters.length) waiters.shift().reject(e); });
  ws.on("close", () => { if (!failed) { failed = new Error("ws closed"); while (waiters.length) waiters.shift().reject(failed); } });
  const next = () => (queue.length ? Promise.resolve(queue.shift()) : failed ? Promise.reject(failed) : new Promise((res, rej) => waiters.push({ resolve: res, reject: rej })));
  await new Promise((res, rej) => { ws.once("open", res); ws.once("error", rej); });

  const sessionId = randomUUID();
  try {
    ws.send(encode({ event: EVENT.StartConnection }));
    for (;;) { const m = await next(); if (m.event === EVENT.ConnectionStarted) break; if (m.type === MSG_TYPE.Error || m.event === EVENT.ConnectionFailed) throw new Error("connect failed: " + m.payload.toString()); }
    ws.send(encode({
      event: EVENT.StartSession,
      sessionId,
      payload: JSON.stringify({
        req_params: {
          model,
          speaker: voice,
          audio_params: { format: "mp3", sample_rate: sampleRate, speech_rate: speechRate, enable_subtitle: false },
          additions: JSON.stringify({ disable_markdown_filter: true, explicit_language: "zh-cn" })
        }
      })
    }));
    for (;;) { const m = await next(); if (m.event === EVENT.SessionStarted) break; if (m.type === MSG_TYPE.Error || m.event === EVENT.SessionFailed) throw new Error("session failed: " + m.payload.toString()); }

    const t0 = Date.now();
    ws.send(encode({ event: EVENT.TaskRequest, sessionId, payload: JSON.stringify({ req_params: { text } }) }));
    ws.send(encode({ event: EVENT.FinishSession, sessionId }));

    const chunks = [];
    let firstPacketMs = -1;
    for (;;) {
      const m = await next();
      if (m.type === MSG_TYPE.Error || m.event === EVENT.SessionFailed) throw new Error("tts failed: " + m.payload.toString());
      if (m.event === EVENT.TTSResponse && m.type === MSG_TYPE.AudioOnlyServer) {
        if (firstPacketMs < 0) firstPacketMs = Date.now() - t0;
        chunks.push(m.payload);
      }
      if (m.event === EVENT.SessionFinished) break;
    }
    ws.send(encode({ event: EVENT.FinishConnection }));
    return { audio: Buffer.concat(chunks), firstPacketMs };
  } finally {
    try { ws.close(); } catch { /* noop */ }
  }
}

File: pipeline/src/saydo_pipeline/doubao_tts.py
"""doubao seed-tts-2.0 v3 双向流式客户端(Python 移植;协议同 e2e/spikes/asr-1.0/doubao-tts.mjs,
上游来源 = owner repo-demo-recorder 实现,07 D5 指定复用)。

无状态、每句一连接(P0 简化;双向流式长连接优化留 P1)。
"""

from __future__ import annotations

import json
import struct
import uuid
from dataclasses import dataclass

import websockets

MSG_FULL_CLIENT = 0b0001
MSG_FULL_SERVER = 0b1001
MSG_AUDIO_ONLY = 0b1011
MSG_ERROR = 0b1111
FLAG_WITH_EVENT = 0b100

EV_START_CONNECTION = 1
EV_FINISH_CONNECTION = 2
EV_CONNECTION_STARTED = 50
EV_CONNECTION_FAILED = 51
EV_START_SESSION = 100
EV_FINISH_SESSION = 102
EV_SESSION_STARTED = 150
EV_SESSION_FINISHED = 152
EV_SESSION_FAILED = 153
EV_TASK_REQUEST = 200
EV_TTS_RESPONSE = 352

_NO_SESSION_EVENTS = {EV_START_CONNECTION, EV_FINISH_CONNECTION, EV_CONNECTION_STARTED, EV_CONNECTION_FAILED, 52}


def _encode(event: int, session_id: str = "", payload: str = "{}") -> bytes:
    header = bytes([0x11, (MSG_FULL_CLIENT << 4) | FLAG_WITH_EVENT, 0x10, 0x00])
    body = struct.pack(">i", event)
    if event not in _NO_SESSION_EVENTS:
        sid = session_id.encode()
        body += struct.pack(">I", len(sid)) + sid
    pb = payload.encode()
    body += struct.pack(">I", len(pb)) + pb
    return header + body


@dataclass
class _Msg:
    type: int
    event: int
    payload: bytes


def _decode(data: bytes) -> _Msg:
    header_size = (data[0] & 0x0F) * 4
    mtype = data[1] >> 4
    flag = data[1] & 0x0F
    off = header_size
    if mtype in (MSG_FULL_CLIENT, MSG_FULL_SERVER, MSG_AUDIO_ONLY):
        if flag in (0b1, 0b11):
            off += 4
    elif mtype == MSG_ERROR:
        off += 4
    event = 0
    if flag == FLAG_WITH_EVENT:
        event = struct.unpack(">i", data[off : off + 4])[0]
        off += 4
        if event not in _NO_SESSION_EVENTS:
            n = struct.unpack(">I", data[off : off + 4])[0]
            off += 4 + n
        if event in (EV_CONNECTION_STARTED, EV_CONNECTION_FAILED, 52):
            n = struct.unpack(">I", data[off : off + 4])[0]
            off += 4 + n
    payload = b""
    if off + 4 <= len(data):
        n = struct.unpack(">I", data[off : off + 4])[0]
        off += 4
        payload = data[off : off + n]
    return _Msg(mtype, event, payload)


class DoubaoTts:
    """每句一连接的合成客户端;synthesize 产出 (mp3_bytes, first_packet_ms)。"""

    def __init__(
        self,
        api_key: str,
        resource_id: str = "seed-tts-2.0",
        model: str = "seed-tts-2.0-expressive",
        voice: str = "zh_female_tianmeiyueyue_uranus_bigtts",
        sample_rate: int = 24000,
        endpoint: str = "wss://openspeech.bytedance.com/api/v3/tts/bidirection",
    ) -> None:
        self.api_key = api_key
        self.resource_id = resource_id
        self.model = model
        self.voice = voice
        self.sample_rate = sample_rate
        self.endpoint = endpoint

    async def synthesize(self, text: str) -> bytes:

        headers = {
            "X-Api-Key": self.api_key,
            "X-Api-Resource-Id": self.resource_id,
            "X-Api-Connect-Id": str(uuid.uuid4()),
        }
        session_id = str(uuid.uuid4())
        async with websockets.connect(self.endpoint, additional_headers=headers, max_size=None, proxy=None) as ws:
            await ws.send(_encode(EV_START_CONNECTION))
            while True:
                m = _decode(await ws.recv())
                if m.event == EV_CONNECTION_STARTED:
                    break
                if m.type == MSG_ERROR or m.event == EV_CONNECTION_FAILED:
                    raise RuntimeError(f"doubao connect failed: {m.payload[:200]!r}")
            await ws.send(
                _encode(
                    EV_START_SESSION,
                    session_id,
                    json.dumps(
                        {
                            "req_params": {
                                "model": self.model,
                                "speaker": self.voice,
                                "audio_params": {
                                    "format": "mp3",
                                    "sample_rate": self.sample_rate,
                                    "enable_subtitle": False,
                                },
                                "additions": json.dumps(
                                    {"disable_markdown_filter": True, "explicit_language": "zh-cn"}
                                ),
                            }
                        }
                    ),
                )
            )
            while True:
                m = _decode(await ws.recv())
                if m.event == EV_SESSION_STARTED:
                    break
                if m.type == MSG_ERROR or m.event == EV_SESSION_FAILED:
                    raise RuntimeError(f"doubao session failed: {m.payload[:200]!r}")
            await ws.send(_encode(EV_TASK_REQUEST, session_id, json.dumps({"req_params": {"text": text}})))
            await ws.send(_encode(EV_FINISH_SESSION, session_id))
            chunks: list[bytes] = []
            while True:
                m = _decode(await ws.recv())
                if m.type == MSG_ERROR or m.event == EV_SESSION_FAILED:
                    raise RuntimeError(f"doubao tts failed: {m.payload[:200]!r}")
                if m.event == EV_TTS_RESPONSE and m.type == MSG_AUDIO_ONLY:
                    chunks.append(m.payload)
                if m.event == EV_SESSION_FINISHED:
                    break
            await ws.send(_encode(EV_FINISH_CONNECTION))
            return b"".join(chunks)
